### DISCUSSION TITLE
Speed up cross-validation

### DISCUSSION THREAD

**Discussion by S90pngprgm** [kogalur/randomForestSRC#209] [Q&A] [answered]
Suppose that I split a data set S into n disjoint subsets, S1, ..., Sn.

Each time I use the whole data set S excluding Si (i.e., S\Si for i = 1, ..., n) as the training data, and predict for Si. A naive solution has a time complexity of n as I need to train a random forest model n times.

But given the similarities among these n models, is there a way to run random forest on this group of subsets S\Si more efficiently? Can the time complexity be reduced to be independent of n?

**Comment by ishwaran** [comment]
The first thing I want to point out is that the out-of-bag estimated values returned by the package are already essentially leave-one-out (LOO) cross-validated prediction.  Technically they are bootstrap LOO, which is actually better because they are smoother and have better variance property.  These LOO values are stored `o$predicted.oob` for every random forest family.

**Comment by ishwaran** [comment, marked as answer]
Next, if you _really_ want traditional cross-validation, then there is a way to do this efficiently in the package.  My suggestion is to use the custom bootstrap option` bootstrap='by.user' `combined with the `samp` option.  Suppose you want 5-fold validation.  Then specify 100 trees for each fold to be left out.  This creates a forest of 500 trees where each block of 100 trees leaves out a fold and therefore over those 100 trees you can obtain out-of-sample predicted values for that fold.  In order to extract the 100 trees for each fold, use the `get.tree` option in prediction which allows you to extract specific trees or groups of trees over which you want performance values from.

Here's the code illustrating this for the `mtcars` data set.   Note that `samp` currently requires all trees to have the same sample size.  So I had to slightly adjust the cross-validation to take that into account.

```
## use mtcars with 100 trees for each fold using 5-fold validation
n <- nrow(mtcars)
K <- 5
ntreeK <- 100

## create cv test fold list
cv <- randomForestSRC:::cv.folds(n, K)

## balance the cv test fold list
N <- min(sapply(cv, length))
cv <- lapply(cv, function(vl) {vl[1:N]})

## create the user specified bootstrap
trn <- lapply(cv, function(vl) {setdiff(1:n, vl)})
samp <- do.call(cbind, lapply(trn, function(vl) {
  vlmtx <- rep(0, n)
  vlmtx[vl] <- 1
  matrix(vlmtx, n, ntreeK)
  })
)

## fit forest using custom bootstrap sample
ntree <- K * ntreeK
o <- rfsrc(mpg ~ ., mtcars, bootstrap = "by.user", samp = samp, ntree = ntree)

## get out-of sample predicted values for each fold
bytree <- c(0, (1:K) * ntreeK)
pred <- lapply(1:K, function(k) {
  yhat <- predict(o, mtcars[cv[[k]],, drop = FALSE],
                  get.tree = (bytree[k]+1):bytree[k+1])$predicted
  y <- mtcars[cv[[k]], "mpg"]
  data.frame(id = cv[[k]], y = y, yhat = yhat)
})

```
